pkg/builder/in_memory_build_queue.go

```diff
@@ -1091,13 +1350,26 @@ func (h queuedOperationsHeap) Len() int {
 	return len(h)
 }
 
-func (h queuedOperationsHeap) Less(i, j int) bool {
-	// Lexicographic order on priority and queued timestamp.
-	ti := h[i].operation.desiredState.QueuedTimestamp
-	tj := h[j].operation.desiredState.QueuedTimestamp
-	return h[i].priority < h[j].priority || (h[i].priority == h[j].priority &&
-		(ti.Seconds < tj.Seconds || (ti.Seconds == tj.Seconds &&
-			ti.Nanos < tj.Nanos)))
+func (h queuedOperationsHeap) Less(x, y int) bool {
+	// Lexicographic order on priority, invocation timestamp and queued timestamp.
+	// The invocation timestamp used is the oldest timestamp for connected invocations.
+	qx := h[x].operation.desiredState.QueuedTimestamp
+	qy := h[y].operation.desiredState.QueuedTimestamp
+
+	return h[x].priority < h[y].priority ||
```

Comment:
I think the readability of this function could be improved if it was written as follows:
```go
if vx, vy := h[x].priority, h[y].priority; vx < vy {
    return true
} else if vx > vy {
    return false
}
if vx, vy := h[x].operation.invocations[0].startTime, h[y].operation.invocations[0].startTime; vx < vy {
    return true
} else if vx > vy {
    return false
}
...
```
This allows us to keep on adding more conditions without blowing up expression size.

That said, are we happy with the selection logic that is used here? It looks like you’re picking the start time of the build to determine which action gets preferred over the other one, right? Doesn’t this mean that long-running builds that have been running for some time will cause starvation for short-running builds that have only been started recently?

My suggestion would be to basically replace queuedOperationsHeap by a different data structure: a heap of heaps.

- The top-level heap would contain every Bazel invocation for which one or more build actions are queued for that platform. This is a min-heap that is ordered by the number of actions that this invocation has running on that platform right now.
- The second-level heap would contain every operation associated with that Bazel invocation. This is a min-heap that is ordered by priority and queueing time.

By using an approach like this, every Bazel client will get its fair share of workers. The scheduler will continuously allow the Bazel client that runs the least number of build actions to run more actions.

Reply:
I like the idea about ordering bazel invocations, rather than actions, in a top-level heap.

But shouldn’t the priority field be more significant? Not having priority present at all in the top-level heap would make priority almost useless. Right?

And if there are several invocations with equal number of actions running, then the oldest invocation should be scheduled first, to avoid starvation. Right?

It seems that there is no single algorithm that fit all use cases and expected load models. That could be a good reason for making it configurable. Would the following give enough flexibility for most use cases?

Top-level heap of bazel invocations ordered by:
1. priority + max(0, (numberOfActionsRunningForTheInvocation - X))
2. invocation start time

where X would be an integer set in bb-schedulers configuration file. Setting X=0 would share capacity equal between invocations, similar to what @EdSchouten suggested. A very high X value would schedule oldest invocation, as this PR is doing. Configuring X to something in between would be a compromise between the two extremes.

I’m not sure of a good name for ‘X’. Maybe ReservedCapacityPerInvocation?

Adding the priority and numberOfActionsRunningForTheInvocation integers together gives flexibility for different use cases, since a priority close to the default 0 would influence the scheduling slightly, while a priority value like -100000 would be very significant.

What do you think @EdSchouten?

Reply:
I think it depends on how 'priority' should be interpreted. It can be interpreted in two different ways:
- There is the concept of high priority vs low priority builds. If two invocations of Bazel are running concurrently, which one should get more workers than the other?
- There is the concept of action-level priority. Given a set of actions being scheduled as part of a single build, which action can best be run first to ensure that the total duration of the build stays close to the critical path?

In case the idea behind REv2's priority field was to cater to that second use case, then the approach that I laid out is exactly what we'd want: treat all builds equal, but let the priority field completely dominate what gets scheduled as part of that build.

In case it's the former, then things become hairy. Especially considering the fact that it's not as if a Bazel invocation has a single priority. It's composed of build actions that *each* have their own priority. It's basically a two-dimensional thing, and we need to turn it into something one-dimensional. Which means I agree with your statement:
> It seems that there is no single algorithm that fit all use cases and expected load models.

That said, if we were to simply make the assumption that Bazel always sends the same priority value for all of its build actions, then I think we can come up with a strategy that requires a bit less magic than what you laid out.

What if the default policy (in case priority == 0 for all builds) is that all builds get exactly the same set of workers (so n/k, where n is the number of workers, k is the number of Bazel builds going on at that point in time). Then, the priority value can be used to influence a weight factor.

The naïve approach would be to use the priority directly as the weight factor through multiplication, but that's annoying. You have to deal with priority being non-positive. This is why you came up with this expression:

> priority + max(0, (numberOfActionsRunningForTheInvocation - X))

Instead of using that, what are your thoughts on using an exponential weight? An awesome property of exponentiation is that the exponent can be an arbitrary number, and the output is always positive (assuming a positive base). So if you set priority to zero, your build gets weight `b^0 = 1`. If you pick priority=1, you get `b` times as many (as `b^1 = b`). It doesn't require any explicitly configured bounds on the scheduler.

Now the question is what `b` should be. Let me propose a value: `b = 2^(1/100) =~ 1.00696`. This means that if you set priority=100, you get twice as many workers as another build running with priority=0. If you set priority=200, you get four times as many. That should give people more than enough granularity.

Reply:
I don’t know about BuildStrem, RECC, etc, but I have not seen support in bazel for priorities per action, only per invocation. Maybe I missed something? If not, is it good enough taking priority from first seen action, as in this PR?

If we would like to honor the granularity in REv2, we could consider letting each invocation in top-level heap, inherit the priority of next queued action, from its associated second-level heap. And/or have a configuration option for ignoring priority in top-level heap?

> This is why you came up with this expression

Yes, you are reading my mind :smiley:

Your idea with `b^priority` seems better than my attempt adding integers.  Do you imagine the weight factor to be used like the following?

`(b^priority) * (1 + numberOfActionsRunningForTheInvocation)`

And if also including a configurable X parameter, could the top-level heap be ordered by the following?

1.	`(b^priority) * (1 + max(0, (numberOfActionsRunningForTheInvocation – X)))`
2.	`invocation start time`

Reply:
I'd say that the top level uses a min-heap, sorted by the expression:
```
number_of_actions_running_for_the_invocation / b^priority
```

Reply:
> I'd say that the top level uses a min-heap, sorted by the expression:
> 
> ```
> number_of_actions_running_for_the_invocation / b^priority
> ```

-100 in REv2 priority field means higher priority than 100. Right?

Higher: `b^-100 =~ 0.5`
Lower: `b^100 =~ 2`

So for a min-heap where we want the higher priority to get the smallest value, we need to multiply, not divide. Right?

`number_of_actions_running_for_the_invocation * b^priority`

And when there are several invocations with zero actions running, their resulting product would all be 0, **regardless of their priority**. To assure they are started in priority order, we can calculate:

`(1 + number_of_actions_running_for_the_invocation) * b^priority`

> Surely we can take the start time into account as well, but I don't think that's really necessary. I don't think that can lead to sustainable starvation of any sort, right? Every time a certain invocation gets an operation assigned to a worker, it'll become harder for it to get more operations assigned, due to number_of_actions_running_for_the_invocation increasing.

That is true when cluster capacity is significantly higher than number of invocations, but not if there are many invocations waiting without any of their actions running.

And if having a configurable X constant in the formula, for my use case, then that secondary ordering by invocation start time would be essential.

Reply:
You're 100% correct!